pilot/pkg/xds/workload.go

```diff
@@ -112,17 +113,25 @@ func (e WorkloadGenerator) GenerateDeltas(
 		}
 	}
 
-	if !w.Wildcard {
-		// For on-demand, we may have requested a VIP but gotten Pod IPs back. We need to update
-		// the internal book-keeping to subscribe to the Pods, so that we push updates to those Pods.
-		w.ResourceNames = sets.SortedList(sets.New(w.ResourceNames...).Merge(have))
-	}
+	removed = sets.SortedList(removeSet)
 	if full {
 		// If it's a full push, AddressInformation won't have info to compute the full set of removals.
 		// Instead, we need can see what resources are missing that we were subscribe to; those were removed.
 		removes := subs.Difference(have).InsertAll(removed...)
+		// TODO: we may not need to sort, as we donot sort for other resource
 		removed = sets.SortedList(removes)
 	}
```

Comment:
can avoid double sorting:

```go
	if full {
		// If it's a full push, AddressInformation won't have info to compute the full set of removals.
		// Instead, we need can see what resources are missing that we were subscribe to; those were removed.
		removed = sets.SortedList(subs.Difference(have).Merge(removeSet))
	} else {
		removed = sets.SortedList(removeSet)
	}
```

Reply:
I want to remove all sortings, not sure any ut will fail.

Reply:
removed sorting